File: pat(advanced)C++/1021(advanced).cpp
#include<iostream>
#include<vector>
#include<set>
using namespace std;

//set本身就有去重功能和排序功能


vector<int>v[10001];
vector<int>farthest_point;
int maxdis = 0;
int visited[10001] = {0};
set<int>final_ans;

void dfs(int cur){
	if(visited[cur]){
		return;//及时止损
	}
	visited[cur] = 1;
	for(int m:v[cur]){
		dfs(m);
	}
}

void dfs2(int cur,int depth){
	if(visited[cur])return;
	visited[cur] = 1;
	if(depth>maxdis) {
		maxdis = depth;
		farthest_point.clear();
		farthest_point.emplace_back(cur);
	}
	else if(depth == maxdis){
		farthest_point.emplace_back(cur);
	}
	for(int i:v[cur]) {
		dfs2(i,depth+1);
	}
}

int main(){
	int N,i,j,k;
	cin>>N;
	for(i = 0;i<N;i++){
		cin>>j>>k;
		v[j].emplace_back(k);
		v[k].emplace_back(j);
	}
	int count = 0;
	for(i = 1;i<=N;i++){
		if(!visited[i]){
			count++;
			dfs(i);
		}
	}
	if(count>1){
		cout<<"Error: "<<count<<" components";
		//不是树
	}
	else{
		//是树
		for(i = 1;i<N;i++){
			visited[i] = 0;
		}
		dfs2(1,0);//第一次dfs2
		for(int each:farthest_point)final_ans.insert(each);
		farthest_point.clear();
		maxdis = 0;
		for(i = 0;i<10001;i++)visited[i] = 0;
		dfs2(*final_ans.begin(),0);//第二次dfs2
		for(int each:farthest_point){
			final_ans.insert(each);
		}
		for(int each:final_ans){
			cout<<each<<'\n';
		}
	}
	return 0;
}

File: pat(advanced)python/1021(advanced).py
import sys


def dfs(node):
    if visited[node - 1] == 1:
        return
    visited[node - 1] = 1
    for i in tables[node - 1]:
        dfs(i)


def dfs2(node, depth):
    global maxdepth
    if (visited1[node - 1] == 1):
        return
    visited1[node - 1] = 1
    for i in tables[node - 1]:
        dfs2(i, depth + 1)
    if maxdepth < depth:
        maxdepth = depth
        middle_answer.clear()
        middle_answer.append(node)
    elif maxdepth == depth:
        middle_answer.append(node)


if __name__ == '__main__':
    number_nodes = int(input())
    if number_nodes == 1:
        print(1)
    else:
        tables = [[] for x in range(number_nodes)]
        for i in range(number_nodes - 1):
            line = list(map(int, input().split(' ')))
            node1 = line[0]
            node2 = line[1]
            tables[node1 - 1].append(node2)
            tables[node2 - 1].append(node1)
        visited = [0 for x in range(number_nodes)]
        count = 0
        for i in range(number_nodes):
            if (visited[i] != 1):
                dfs(i + 1)
                count += 1
        if (count > 1):
            # 不是树
            print('Error: %d components' % count)
        else:
            # 是树
            middle_answer = []
            maxdepth = 0
            visited1 = [0 for j in range(number_nodes)]
            dfs2(1, 0)
            final_answer = set(middle_answer)
            visited1 = [0 for j in range(number_nodes)]
            maxdepth = 0
            indexnode = middle_answer[0]
            middle_answer = []
            dfs2(indexnode, 0)
            final_answer = list(final_answer | set(middle_answer))
            final_answer.sort()
            for k in final_answer:
                print(k)
